### PR TITLE
[Feature] Support ConvNext

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation

Support ConvNext. 
Paper: https://arxiv.org/abs/2201.03545
Github: https://github.com/facebookresearch/ConvNeXt

## Modification

Need some refactor. WIP.

- [x] Refactor backbone code.
- [x] Refactor `mmcv_custome` code and align inference metric.
- [x] Add unit test
- [x] Align training metric.

### CODE REVIEW

**Comment by MengzhangLI** [comment]
In our refactor code, (NCHW -> permute to NHWC -> PyTorch LN -> linear layers -> layer scale -> permute back to NCHW) takes over (NCHW -> custom LN -> 1x1 convs -> layer scale), the former style may be faster.

Reference: https://github.com/facebookresearch/ConvNeXt/issues/18.

**Review by xvjiarui — COMMENTED** [review]
Please refer to ResNet for some details

File: mmseg/models/backbones/convnext.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+        def _init_weights(m):
+            if isinstance(m, nn.Linear):
+                trunc_normal_(m.weight, std=.02)
+                if isinstance(m, nn.Linear) and m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+            elif isinstance(m, nn.Conv2d):
+                trunc_normal_(m.weight, std=.02)
+                nn.init.constant_(m.bias, 0)
+            elif isinstance(m, nn.LayerNorm):
```
[the remark this answers is not in the document]
Comment (MengzhangLI) [reply mmseg/models/backbones/convnext.py:159]: The value of constant initialization is 0 not 1. In ConvNext original repo, it is `nn.init.constant_(m.bias, 0)`, but our default constant init, `constant_init(m, val=1.0, bias=0.)`.

File: mmseg/models/backbones/convnext.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 in_chans=3,
+                 depths=[3, 3, 9, 3],
+                 dims=[96, 192, 384, 768],
+                 drop_path_rate=0.,
+                 layer_scale_init_value=1e-6,
+                 out_indices=[0, 1, 2, 3],
```
[the remark this answers is not in the document]
Comment (MengzhangLI) [reply mmseg/models/backbones/convnext.py:112]: Does that mean the range of `out_indices`? For example, `out_indices=[0, 1, 2, 99]` is wrong because on 100-th feature maps.

**Comment by mzr1996** [comment]
Please look at https://github.com/open-mmlab/mmclassification/pull/670

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            params (list[dict]): A list of param groups, it will be modified
+                in place.
+            module (nn.Module): The module to be added.
+            prefix (str): The prefix of the module
+            is_dcn_module (int|float|None): If the current module is a
+                submodule of DCN, `is_dcn_module` will be passed to
+                control conv_offset layer's learning rate. Defaults to None.
+        """
+        parameter_groups = {}
+        print(self.paramwise_cfg)
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:99]: use logger.info instead of print

File: configs/convnext/upernet_convnext_base_640x640_160k_ade20k.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+model = dict(
+    backbone=dict(
+        type='mmcls.ConvNeXt',
+        arch='base',
+        out_indices=[0, 1, 2, 3],
+        drop_path_rate=0.4,
+        layer_scale_init_value=1.0,
+        gap_before_final_norm=False,
+        init_cfg=dict(type='Pretrained', checkpoint=checkpoint_file),
+        _delete_=True),
```
Comment (MeowZheng) [inline configs/convnext/upernet_convnext_base_640x640_160k_ade20k.py:17]: `_delete_` might delete all keys in base cfg

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        logger = get_root_logger()
+
+        parameter_groups = {}
+        logger.info(f'self.paramwise_cfg is {self.paramwise_cfg}')
+        num_layers = self.paramwise_cfg.get('num_layers') + 2
+        decay_rate = self.paramwise_cfg.get('decay_rate')
+        decay_type = self.paramwise_cfg.get('decay_type', 'layer_wise')
+        logger.info(
+            f'Build LearningRateDecayOptimizerConstructor {decay_type} {decay_rate} - {num_layers}'  # noqa
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:108]: multi-line string can be concatenated

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                this_weight_decay = weight_decay
+
+            if decay_type == 'layer_wise':
+                layer_id = get_num_layer_layer_wise(
+                    name, self.paramwise_cfg.get('num_layers'))
+                logger.info(f'set param {name} as id {layer_id}')
+            elif decay_type == 'stage_wise':
+                layer_id = get_num_layer_stage_wise(name, num_layers)
+                logger.info(f'set param {name} as id {layer_id}')
+            group_name = 'layer_%d_%s' % (layer_id, group_name)
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:130]: f string

File: requirements/mminstall.txt [hunk 7]
```diff
@@ -1 +1,2 @@
-mmcv-full>=1.3.1,<=1.4.0
+mmcls>=0.20.1,<=0.21
```
Comment (MeowZheng) [inline requirements/mminstall.txt:1]: it might be too narrow

File: requirements/runtime.txt [hunk 8]
```diff
@@ -1,4 +1,5 @@
 matplotlib
+mmcls
```
Comment (MeowZheng) [inline requirements/runtime.txt:2]: mmcls might be >= the version that has `ConvNext`

**Review by MeowZheng — COMMENTED** [review]
might add UTs for optimizer constructor

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+
+def get_num_layer_layer_wise(var_name, num_max_layer=12):
+    """Get the layer id to set the different learning rates in `layer_wise`
+    decay_type.
+
+    Args:
+        var_name (str): The key of the model.
+        num_max_layer (int): Maximum number of backbone layers.
+    """
```
[the remark this answers is not in the document]
Comment (MengzhangLI) [reply mmseg/core/utils/layer_decay_optimizer_constructor.py:17]: OK. By the way, I think I should tell you return value should also be added for the counterpart in BEiT `get_num_layer_for_vit`.

File: tests/test_core/test_learning_rate_decay_optimizer_constructor.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    assert len(param_groups) == len(set(group_name_lst))
+
+
+def test_convnext_learning_rate_decay_optimizer_constructor():
+
+    # paramwise_cfg with ConvNeXtExampleModel
+    model = ConvNeXtExampleModel()
+    optimizer_cfg = dict(
+        type='AdamW', lr=0.0001, betas=(0.9, 0.999), weight_decay=0.05)
+    paramwise_cfg = dict(decay_rate=0.9, decay_type='stage_wise', num_layers=6)
```
Comment (MeowZheng) [inline tests/test_core/test_learning_rate_decay_optimizer_constructor.py:90]: might also check  'layer_wise'?

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        return stage_id + 1
+    else:
+        return num_max_layer - 1
+
+
+@OPTIMIZER_BUILDERS.register_module()
+class LearningRateDecayOptimizerConstructor(DefaultOptimizerConstructor):
+    """Different learning rates are set for different layers of backbone."""
+
+    def add_params(self, params, module, is_dcn_module=None):
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:106]: `is_dcn_module` might not be uesed?

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 12]
```diff
@@ -47,17 +47,6 @@ def get_num_layer_layer_wise(var_name, num_max_layer=12):
         elif stage_id == 2:
             layer_id = num_max_layer
         return layer_id
-    elif var_name.startswith('backbone.downsample_layers'):
```
Comment (MengzhangLI) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py]: L50-60 is redundant with L26-36.

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def get_num_layer_stage_wise(var_name, num_max_layer):
+    """Get the layer id to set the different learning rates in ``stage_wise``
+    decay_type.
+
+    Args:
+        var_name (str): The key of the model.
+        num_max_layer (int): Maximum number of backbone layers.
+    """
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:73]: might add returns description

File: tests/test_core/test_learning_rate_decay_optimizer_constructor.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        self.decode_head = UPerHead(
+            in_channels=[128, 256, 512, 1024],
+            in_index=[0, 1, 2, 3],
+            pool_scales=(1, 2, 3, 6),
+            channels=512,
+            dropout_ratio=0.1,
+            num_classes=19)
+
+    def forward(self, x):
+        return x
```
Comment (MeowZheng) [inline tests/test_core/test_learning_rate_decay_optimizer_constructor.py:45]:
```suggestion

```

File: tests/test_core/test_learning_rate_decay_optimizer_constructor.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        self.backbone.downsample_norm0 = nn.BatchNorm2d(2)
+        self.backbone.downsample_norm1 = nn.BatchNorm2d(2)
+        self.backbone.downsample_norm2 = nn.BatchNorm2d(2)
+        self.backbone.lin = nn.Parameter(torch.ones(1))
+        self.backbone.lin.requires_grad = False
+
+        self.decode_head = UPerHead(
+            in_channels=[128, 256, 512, 1024],
+            in_index=[0, 1, 2, 3],
+            pool_scales=(1, 2, 3, 6),
```
Comment (MeowZheng) [inline tests/test_core/test_learning_rate_decay_optimizer_constructor.py:39]: it might not be necessary to test a such big model

**Review by MeowZheng — COMMENTED** [review]
might add metafile and readme

File: tests/test_core/test_learning_rate_decay_optimizer_constructor.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+            this_weight_decay = weight_decay
+
+        if paramwise_cfg['decay_type'] == 'layer_wise':
+            layer_id = get_num_layer_layer_wise(example_param_name,
+                                                paramwise_cfg['num_layers'])
+        elif paramwise_cfg['decay_type'] == 'stage_wise':
+            layer_id = get_num_layer_stage_wise(example_param_name, num_layers)
+        group_name = f'layer_{layer_id}_{group_name}'
+
+        scale = paramwise_cfg['decay_rate']**(num_layers - layer_id - 1)
```
Comment (MeowZheng) [inline tests/test_core/test_learning_rate_decay_optimizer_constructor.py:81]: might give a ground truth scale

File: README_zh-CN.md [hunk 17]
```diff
@@ -119,6 +119,7 @@ MMSegmentation 是一个基于 PyTorch 的语义分割开源工具箱。它是 O
 - [x] [DPT (ArXiv'2021)](configs/dpt)
 - [x] [Segmenter (ICCV'2021)](configs/segmenter)
 - [x] [SegFormer (NeurIPS'2021)](configs/segformer)
+- [x] [ConvNeXt (ArXiv'2022)](configs/convnext)
```
Comment (MengzhangLI) [inline README_zh-CN.md:122]:
```suggestion

```

File: README.md [hunk 18]
```diff
@@ -120,6 +120,7 @@ Supported methods:
 - [x] [DPT (ArXiv'2021)](configs/dpt)
 - [x] [Segmenter (ICCV'2021)](configs/segmenter)
 - [x] [SegFormer (NeurIPS'2021)](configs/segformer)
+- [x] [ConvNeXt (ArXiv'2022)](configs/convnext)
```
Comment (MengzhangLI) [inline README.md:123]:
```suggestion

```

Comment (MengzhangLI) [reply README.md:123]: Should be set in `Backbones`, not `Methods.`

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    elif var_name.startswith('backbone.downsample_norm'):
+        stage_id = int(
+            var_name.split('backbone.downsample_norm')[1].split('.')[0])
+        if stage_id == 0:
+            layer_id = 2
+        elif stage_id == 1:
+            layer_id = 3
+        elif stage_id == 2:
+            layer_id = num_max_layer
+        return layer_id
```
Comment (MengzhangLI) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:49]: `backbone.stem_norm` and `backbone.downsample_norm` may not needed since they are not existed in mmcls convnext backbone. Same as in `get_num_layer_stage_wise()`.

Comment (MeowZheng) [reply mmseg/core/utils/layer_decay_optimizer_constructor.py:49]: why did add 'if conditions' for `backbone.stem_norm` and `backbone.downsample_norm`,
from convnext official repo, there is not anything about these var_names.
https://github.com/facebookresearch/ConvNeXt/blob/3d444184dc27156e0562133c3b69b56f7efba500/semantic_segmentation/mmcv_custom/layer_decay_optimizer_constructor.py#L14-L42

File: tests/test_core/test_learning_rate_decay_optimizer_constructor.py [hunk 20]
```diff
@@ -0,0 +1,186 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+
+from mmseg.core.utils.layer_decay_optimizer_constructor import \
+    LearningRateDecayOptimizerConstructor
+
+base_lr = 1000000000
+decay_rate = 9
```
Comment (MeowZheng) [inline tests/test_core/test_learning_rate_decay_optimizer_constructor.py:10]: I think 2 might be better for `decay_rate`.

My suggestion is that base_lr might be `1`, and `lr_scale` will equal to `lr` and check it after line 165.

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+def get_num_layer_stage_wise(var_name, num_max_layer):
+    """Get the layer id to set the different learning rates in ``stage_wise``
+    decay_type.
+
+    Args:
+        var_name (str): The key of the model.
+        num_max_layer (int): Maximum number of backbone layers.
+    Returns:
+        int: The id number corresponding to different
+        learning rate in ``LearningRateDecayOptimizerConstructor``.
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:62]:
```suggestion
        int: The id number corresponding to different learning rate in
        ``LearningRateDecayOptimizerConstructor``.
```

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+    """Get the layer id to set the different learning rates in ``layer_wise``
+    decay_type.
+
+    Args:
+        var_name (str): The key of the model.
+        num_max_layer (int): Maximum number of backbone layers.
+
+    Returns:
+        int: The id number corresponding to different
+        learning rate in ``LearningRateDecayOptimizerConstructor``.
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:20]:
```suggestion
        int: The id number corresponding to different learning rate in
        ``LearningRateDecayOptimizerConstructor``.
```

**Review by MeowZheng — APPROVED** [review]
